### PR TITLE
[WIP] feature(application): Introduces \Elgg\Application ...

### PR DESCRIPTION
**Pull request by mrclay**
a continuation on #7939. cc @Srokap 
- [ ] Approval of overall design
- [ ] Agreement of visibility of APIs
- [ ] Mention in docs?
- [ ] Recut into more structured commits

### CODE REVIEW

File: engine/classes/Elgg/Application.php [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+	 * @var string
+	 */
+	private $install_dir;
+
+	/**
+	 * @var string
+	 */
+	private $config_file;
+
+	public function __construct($config_file = '') {
```
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: We might want to just allow injecting config totally here (no need to load config file), but this is still useful I think.

**Comment by mrclay** [comment]
At this point only commit messages are failing!

File: autoloader.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?php
+/**
+ * We handle here two possible locations of composer-generated autoload file. One is directly in Elgg directory
+ * when not using as dependency, other is in the main project dir, parent to this one.
+ */
+
+$paths = [
+	__DIR__ . '/vendor/autoload.php',
+	__DIR__ . '/../../../vendor/autoload.php',
```
Comment (ewinslow) [inline autoloader.php]: I look forward to the day we can just delete this...

File: engine/classes/Elgg/Application.php [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+		// we only depend on it to be defining _elgg_services function
+		require_once "$lib_dir/autoloader.php";
+
+		// set up autoloading and DIC
+		$services = _elgg_services($this);
+
+		// load the rest of the library files from engine/lib/
+		// All on separate lines to make diffs easy to read + make it apparent how much
+		// we're actually loading on every page (Hint: it's too much).
+		$lib_files = array(
```
Comment (ewinslow) [inline engine/classes/Elgg/Application.php]: Should we just load these via composer? https://getcomposer.org/doc/04-schema.md#files

File: engine/start.php [hunk 4]
```diff
@@ -1,90 +1,9 @@
...
-elgg_trigger_event('plugins_boot', 'system');
-
-// Complete the boot process for both engine and plugins
-elgg_trigger_event('init', 'system');
-
-$CONFIG->boot_complete = true;
-
-// System loaded and ready
-elgg_trigger_event('ready', 'system');
+elgg_deprecated_notice('You should load the core using \Elgg\Application::bootCore() instead of including start.php', "2.0.0");
```
Comment (ewinslow) [inline engine/start.php]: Why deprecation notice? Just pull the file since breaking changes.

File: engine/classes/Elgg/Application.php [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+	}
+
+	/**
+	 * Get a singleton instance for testing
+	 *
+	 * @return \Elgg\Application
+	 * @access private
+	 * @internal temporary solution for unit testing
+	 */
+	public static function getTestInstance() {
```
Comment (ewinslow) [inline engine/classes/Elgg/Application.php]: Why are these needed?

**Comment by ewinslow** [comment]
Fix the messages and just ship it!!

File: engine/classes/Elgg/Application.php (hunk 5 above)
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: These were added a bit hastily to get phpunits working. There are a few tests that need access to the application that's created in the bootstrap. There's probably a better way

**Comment by ewinslow** [comment]
Sounds fine. As I said a couple of days ago, this is a net improvement. While I agree that improvements can still be made, I don't want this caught up in review and then blocked on a bunch of conflicts.

Let's just get it in...

**Comment by mrclay** [comment]
Latest commit recomposes objects in more sensical way: Application needs ServiceProvider needs Config.

Config: settings loading is moved here. Setting to specify settings file location or if one should *not* be loaded. Option to not bind to the global $CONFIG. If used, settings loaded from file will be copied into Config's storage. Maybe this is too much experimentation here.

File: engine/classes/Elgg/Application.php (hunk 3 above)
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: We don't want to include them on every request.

**Comment by mrclay** [comment]
Q: should settings.php be loaded before loadCore() or, as it is now, before bootCore()? If moved to loadCore(), then we wouldn't need manual loadSettings() calls to make sure settings have been read.

File: engine/classes/Elgg/Application.php (hunk 3 above)
Comment (Srokap) [inline engine/classes/Elgg/Application.php]: I see no benefit and main gain here is to have more control over what gets loaded when.

File: autoloader.php (hunk 2 above)
Comment (Srokap) [inline autoloader.php]: I wanted to avoid having this additional autoloader.php, but that sees to be the way other packages handle 2 scenarios of autoloading as well.

File: engine/start.php (hunk 4 above)
Comment (Srokap) [inline engine/start.php]: We want folks to start using \Elgg\Application with more granularity instead of including old start.php with side effects, right?

**Comment by Srokap** [comment]
I was seeing some cookie related issues last time I've checked. Let me check it again before merging.

**Comment by mrclay** [comment]
We should also figure out how much of the Application (and its dependencies) will be public API. From Application the user can now get to services directly. Should we bake a couple scenarios into static methods on Application and leave the dynamic methods private?

**Comment by Srokap** [comment]
I imagined that for now services would be the private thing. I expected that public would be 3 stages of bootstrapping (`loadCore`, `bootCore`, `run`) and configuration setup (imagined it a bit differently). I need to still look over current state and think more about it. I'd prefer to not have any static methods. Should be no reason for them.

File: engine/classes/Elgg/Application.php [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+		$events->trigger('init', 'system');
+
+		$config->set('boot_complete', true);
+
+		// System loaded and ready
+		$events->trigger('ready', 'system');
+	}
+
+	/**
+	 * Get the database instance (also loads settings if not yet loaded)
```
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: I added this because devs are currently crudely reading settings directly and using ext/mysql (fails to support multi-DB config files). This makes it easy to get a DB connection w/o loading core yet still inject the full logger during full boot.

Comment (mrclay) [inline engine/classes/Elgg/Application.php]: E.g. we now use this in the CacheHandler instead of constructing objects there.

File: engine/classes/Elgg/Config.php [hunk 7]
```diff
@@ -119,12 +141,24 @@ function get($name, $site_guid = 0) {
...
+
+	/**
+	 * Get a config value for the current site if it's already loaded. This should be used instead of
+	 * reading directly from global $CONFIG.
+	 *
+	 * @param string $name Name of the configuration value
+	 *
+	 * @return mixed Returns null if value isn't set
+	 */
+	public function getVolatile($name) {
```
Comment (mrclay) [inline engine/classes/Elgg/Config.php]: I'm open to suggestions for a better name for this... I also considered adding another argument to get() to do this, but that seemed more messy.

File: engine/classes/Elgg/Application.php (hunk 3 above)
Comment (ewinslow) [inline engine/classes/Elgg/Application.php]: OK. I was thinking that if they are just going to get refactored to thin wrappers around OO services and then ultimately deleted when we go full DI, it seems like there almost no cost to just always loading them, and it simplifies our code a bit.

Comment (ewinslow) [inline engine/classes/Elgg/Application.php]: It also would give us concrete motivation to continue the refactoring.

**Comment by mrclay** [comment]
Replaced by #8247
